ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,534 @@
+  static final String FULL_ACCOUNT_METADATA_PATH = "/account_metadata/full_data";
+  private static final String ZN_RECORD_ID = "full_account_metadata";
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final HelixPropertyStore<ZNRecord> helixStore;
+  private final Cache cache = new Cache();
+  private volatile boolean isOpen = false;
+
+  /**
+   * Constructor. It fetches the full account metadata set and caches locally.
+   * This call is blocking until it fetches all the {@link Account} metadata from {@link HelixPropertyStore}.
+   *
+   * @param storeConfig The config needed to start a {@link HelixPropertyStore}. Cannot be {@code null}.
+   * @param storeFactory The factory to generate a {@link HelixPropertyStore}. Cannot be {@code null}.
+   */
+  public HelixAccountService(HelixPropertyStoreConfig storeConfig, HelixPropertyStoreFactory<ZNRecord> storeFactory) {
+    if (storeConfig == null || storeFactory == null) {
+      throw new IllegalArgumentException(
+          "storeConfig=" + storeConfig + " and storeFactory=" + storeFactory + " cannot be null");
+    }
+    long startTimeMs = System.currentTimeMillis();
+    logger.info("Starting HelixAccountService with HelixPropertyStore host={} rootPath={}",
+        storeConfig.zkClientConnectString, storeConfig.rootPath);
+    helixStore = storeFactory.getHelixPropertyStore(storeConfig, null);
+    readFullAccountAndUpdateCache(FULL_ACCOUNT_METADATA_PATH);
+    isOpen = true;
+    logger.info("HelixAccountService started. {} accounts have been loaded, took time={}ms",
+        cache.getAccountTwoKeyMap().getAccounts().size(), System.currentTimeMillis() - startTimeMs);
+  }
+
+  @Override
+  public Account getAccountByName(String accountName) {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+    if (accountName == null) {
+      throw new IllegalArgumentException("accountName cannot be null.");
+    }
+    return cache.getAccountByName(accountName);
+  }
+
+  @Override
+  public Account getAccountById(short id) {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+    return cache.getAccountById(id);
+  }
+
+  @Override
+  public Collection<Account> getAllAccounts() {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+    return cache.getAllAccounts();
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p>
+   *   This call is blocking until it completes the operation of updating account metadata to {@link HelixPropertyStore}.
+   * </p>
+   * <p>
+   *   There is a slight chance that an {@link Account} could be updated successfully, even if its value was
+   *   set based on an outdated {@link Account}. This can be fixed after {@code generationId} is introduced
+   *   to {@link Account}.
+   * </p>
+   */
+  @Override
+  public boolean updateAccounts(Collection<Account> accounts) {
```

Comment:
just to record offline discussion. Can you clarify what would happen in this case. Lets say property store is empty and two threads are trying to update with (1, a) and (2,a). If these two threads are made sequential, 2nd call should have failed. But incase of multi-threaded, where exactly the failure would happen.